arrabbiata/src/witness.rs

```diff
@@ -252,6 +281,93 @@ where
             .map(|(l, r)| l + &r.scale(chal))
             .collect();
     }
+
+    pub fn compute_cross_terms(
+        &mut self,
+        witness: &[Vec<BigInt>],
+        challenges: &Challenges<BigInt>,
+        domain_size: usize,
+        gadgets: &[Gadget],
+        constraints: &HashMap<Gadget, Vec<Sparse<Fp, { MV_POLYNOMIAL_ARITY }, { MAX_DEGREE }>>>,
+    ) {
+        (0..domain_size).for_each(|row| {
+            // wrap around if necessary
+            let next_row = (row + 1) % domain_size;
+            let state_left: [E1::ScalarField; NUMBER_OF_COLUMNS * 2] = std::array::from_fn(|i| {
+                if i < NUMBER_OF_COLUMNS {
+                    self.accumulated_program_state[i][row]
+                } else {
+                    self.accumulated_program_state[i - NUMBER_OF_COLUMNS][next_row]
+                }
+            });
+            let state_right: [E1::ScalarField; NUMBER_OF_COLUMNS * 2] = std::array::from_fn(|i| {
+                if i < NUMBER_OF_COLUMNS {
+                    let v: BigUint = witness[i][row].to_biguint().unwrap();
+                    E1::ScalarField::from_biguint(&v).unwrap()
+                } else {
+                    let v: BigUint = witness[i - NUMBER_OF_COLUMNS][row].to_biguint().unwrap();
+                    E1::ScalarField::from_biguint(&v).unwrap()
+                }
+            });
+            let alpha_left = {
+                let v: BigUint = self.accumulated_challenges[ChallengeTerm::ConstraintCombiner]
+                    .to_biguint()
+                    .unwrap();
+                E1::ScalarField::from_biguint(&v).unwrap()
+            };
+            let alpha_right = {
+                let v: BigUint = challenges[ChallengeTerm::ConstraintCombiner]
+                    .to_biguint()
+                    .unwrap();
+                E1::ScalarField::from_biguint(&v).unwrap()
+            };
+            let u_left = {
+                let v: BigUint = self.accumulated_challenges[ChallengeTerm::ConstraintHomogeniser]
+                    .to_biguint()
+                    .unwrap();
+                E1::ScalarField::from_biguint(&v).unwrap()
+            };
+            let u_right = {
+                let v: BigUint = challenges[ChallengeTerm::ConstraintHomogeniser]
+                    .to_biguint()
+                    .unwrap();
+                E1::ScalarField::from_biguint(&v).unwrap()
+            };
+            let activated_gadget = gadgets[row];
```

Comment:
This is not a good approach. We should rely on selectors. The terms "selector" and "gadget" are cofounded because there is a one-to-one correspondence for now.

Bad design! Thinking about changing it while writing an example app like in #3061